libs/deer/src/impls/core/integral.rs

```diff
@@ -1,65 +1,246 @@
-                de: D,
-            ) -> error_stack::Result<Self, DeserializeError> {
-                de.$deser($name).change_context(DeserializeError)
-            }
+macro_rules! num_try_from {
+    ($primitive:ident:: $visit:ident) => {
+        fn $visit(self, v: $primitive) -> Result<Self::Value, VisitorError> {
+            Self::Value::try_from(v)
+                .into_report()
+                .change_context(ValueError.into_error())
+                .attach(ExpectedType::new(self.expecting()))
+                .attach(ReceivedValue::new(v))
+                .change_context(VisitorError)
         }
     };
+}
 
-    ($($typ:ident:: $fun:ident() <- $name:ident. $visit:ident()),*$(,)?) => {
-        $(impl_integral!(@num $name, $visit, $fun, $typ);)*
+macro_rules! num_number {
+    ($primitive:ident:: $to:ident) => {
+        fn visit_number(self, v: Number) -> Result<Self::Value, VisitorError> {
+            v.$to().ok_or_else(|| {
+                Report::new(ValueError.into_error())
+                    .attach(ExpectedType::new(self.expecting()))
+                    .attach(ReceivedValue::new(v))
+                    .change_context(VisitorError)
+            })
+        }
     };
 }
 
-// TODO: fit smaller values (visit) into them / try to fit them
-impl_integral![
-    u8::deserialize_u8() <- U8Visitor.visit_u8(),
-    u16::deserialize_u16() <- U16Visitor.visit_u16(),
-    u32::deserialize_u32() <- U32Visitor.visit_u32(),
-    u64::deserialize_u64() <- U64Visitor.visit_u64(),
-    u128::deserialize_u128() <- U128Visitor.visit_u128(),
-    usize::deserialize_usize() <- USizeVisitor.visit_usize(),
-];
-
-impl_integral![
-    i8::deserialize_i8() <- I8Visitor.visit_i8(),
-    i16::deserialize_i16() <- I16Visitor.visit_i16(),
-    i32::deserialize_i32() <- I32Visitor.visit_i32(),
-    i64::deserialize_i64() <- I64Visitor.visit_i64(),
-    i128::deserialize_i128() <- I128Visitor.visit_i128(),
-    isize::deserialize_isize() <- ISizeVisitor.visit_isize(),
-];
+impl_num!(
+    u8::deserialize_u8;
+    num_self!(u8::visit_u8);
+    num_try_from!(i8::visit_i8, i16::visit_i16, i32::visit_i32, i64::visit_i64, i128::visit_i128, u16::visit_u16, u32::visit_u32, u64::visit_u64, u128::visit_u128);
+    num_number!(u8::to_u8);
+);
+
+impl_num!(
+    u16::deserialize_u16;
+    num_self!(u16::visit_u16);
+    num_from!(u8::visit_u8);
+    num_try_from!(i8::visit_i8, i16::visit_i16, i32::visit_i32, i64::visit_i64, i128::visit_i128, u32::visit_u32, u64::visit_u64, u128::visit_u128);
+    num_number!(u16::to_u16);
+);
+
+impl_num!(
+    u32::deserialize_u32;
+    num_self!(u32::visit_u32);
+    num_from!(u8::visit_u8, u16::visit_u16);
+    num_try_from!(i8::visit_i8, i16::visit_i16, i32::visit_i32, i64::visit_i64, i128::visit_i128, u64::visit_u64, u128::visit_u128);
+    num_number!(u32::to_u32);
+);
+
+impl_num!(
+    u64::deserialize_u64;
+    num_self!(u64::visit_u64);
+    num_from!(u16::visit_u16, u32::visit_u32);
```

Comment:
Should this have `u8::visit_u8` in `num_from`?

Reply:
You are completely right, don't know I missed that.